### PR TITLE
feat: Add support for a generic testing interface

### PR DESCRIPTION
**Pull request by tystuyfzand**
This adds support for other testing libraries that use interfaces like `testing.T` to run tests. I have run `go mod tidy` which may have cleaned up references here, though the changes appear minimal and arikawa has been updated to rc.3

Specific examples of existing implementations of this are:

- [stretchr/testify](https://github.com/stretchr/testify/blob/1e36bfe10404cb77c12f6dfc8665564f3a41ad7e/assert/assertions.go#L28-L30)
- [ginkgo](https://onsi.github.io/ginkgo/#third-party-integrations)

### CODE REVIEW

File: internal/testing/testing.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package testing
+
+// TInterface is a *testing.T interface for using ginkgo/etc with dismock
```
Comment (mavolin) [inline internal/testing/testing.go]:
```suggestion
// TInterface is a *testing.T interface for using ginkgo/etc with dismock.
```

Add a period at the end of the sentence

File: pkg/dismock/dismock_test.go [hunk 2]
```diff
@@ -8,6 +8,7 @@ import (
 
 	"github.com/diamondburned/arikawa/v3/api"
 	"github.com/diamondburned/arikawa/v3/discord"
+	dismockTesting "github.com/mavolin/dismock/v3/internal/testing"
```
Comment (mavolin) [inline pkg/dismock/dismock_test.go]: Don't use camelCase here, as Effective Go discourages it:
> there should be no need for underscores or mixedCaps

Instead either lowercase it, of, if you find that `dismocktesting` is to hard to read, use a shorter alias like `testingutil` or `testutil`.

**Review by mavolin — CHANGES_REQUESTED** [review]
Thank you for your PR. I'm ready to merge, once these two (small) changes are resolved.

The CI is currently failing, because `testing.T.Setenv` is only available since Go 1.17 (at least I think it was added with Go 1.17) and the GitHub setup-go action uses Go 1.15 because of caching. It would be great if you could add the `check-latest: true` field to the workflows, c.f. https://github.com/actions/setup-go#check-latest-version . That should get the tests running.

Since Go 1.18 is soon to be released, I don't think it is necessary to make a build-tagged file that only builds for Go 1.17 higher and includes `Setenv`, and one file that doesn't.

**Comment by tystuyfzand** [comment]
Updated! Let me know if that fails due to the lack of version definition, hopefully check-latest works by itself.

We could also remove Setenv, and trim down the interface significantly if we wanted to in order to only require what we need, but I think using a relatively new version of Go (especially with 1.18 around the corner) it'd be good to keep as-is.

**Review by mavolin — APPROVED** [review]

**Comment by mavolin** [comment]
The codecov action still uses the old Go version, could you change that too?

**Comment by tystuyfzand** [comment]
Should be fixed now, Github is quite picky with actions I guess

**Comment by tystuyfzand** [comment]
Looks like they failed again, let me do some more research into actions and make sure I get this fixed properly.

Edit: or... not?

**Comment by mavolin** [comment]
Seems like it still uses Go 1.15 for whatever reason. Just remove the check latest option and define 1.17 as the Go version.

**Comment by mavolin** [comment]
I can fix it/look into why check-latest does not work myself tomorrow if you prefer.

**Comment by tystuyfzand** [comment]
@mavolin 

It was because I didn't put it in the setup-go section, instead I put it in the "run" section. I need to RTFM before doing changes like that and not rush it...

**Comment by tystuyfzand** [comment]
Looks like it's working now, hopefully you can squash these commits so it doesn't look as bad in the commit history... I should've enabled actions on my fork and made sure it worked.

**Comment by mavolin** [comment]
No worries, it's working now and is merged, and at the end of the day that is all that matters.

**Pull request merged**